lib/SILOptimizer/Transforms/AllocBoxToStack.cpp

```diff
@@ -337,15 +378,32 @@ findUnexpectedBoxUse(SILValue Box, bool examinePartialApply,
       continue;
     }
 
-    // For partial_apply, if we've been asked to examine the body, the
-    // uses of the argument are okay there, and the partial_apply
-    // itself cannot escape, then everything is fine.
-    if (auto *PAI = dyn_cast<PartialApplyInst>(User))
-      if (examinePartialApply && checkPartialApplyBody(Op) &&
-          !partialApplyEscapes(PAI, /* examineApply = */ true)) {
-        LocalPromotedOperands.push_back(Op);
-        continue;
+    if (auto Apply = ApplySite::isa(User)) {
+      if (CurrentRecurDepth > MaxLocalApplyRecurDepth) {
+        return User;
+      }
+      switch (Apply.getKind()) {
+      case ApplySiteKind::PartialApplyInst: {
+        if (checkLocalApplyBody(Op, LocalPromotedOperands, VisitedCallees,
+                                CurrentRecurDepth) &&
+            !partialApplyEscapes(cast<PartialApplyInst>(User),
+                                 /* examineApply = */ true)) {
+          LocalPromotedOperands.push_back(Op);
+          continue;
+        }
+        break;
       }
+      case ApplySiteKind::ApplyInst:
+      case ApplySiteKind::BeginApplyInst:
+      case ApplySiteKind::TryApplyInst:
+        if (isEligibleApply(Apply) &&
```

Comment:
Question, shouldn't isEligibleApply really be isOptimizableLocalApply? Anyways, just a random thought. Feel free to change it in a follow on commit if it makes sense to you. Otherwise, I am ok with this.